### PR TITLE
vector

### PR DESCRIPTION
**Pull request by algebravic**
- add mp-spdz prototype, data generator, and readme
- add IPA_VERBOSE and IPA_NUMROWS_POWER as env variables to set at compile time
- remove using Public Input for row number, did not work
- use exec to unroll loops
- update readme to reflect updates to MP-SPDZ
- iron out final bugs
- add gitattributes for syntax highlighting for .mpc files
- rename _list to reports
- rename ambiguous variables, add Enum for Matrix columns
- Rewriting Erik's ipae2e.mpc using vectors instead of
- Rewrote Erik's code using vectors

### CODE REVIEW

**Comment by bmcase** [comment]
Since this isn't doing a sort, it is really the create helperbits and oblivious attribution stages.   

The great news is the vector operations do seem to be really helping with the compile memory usage.  Comparing to just the oblivious attribution stage before this is using around 13x less memory.

File: research-prototype/Programs/Source/erik.mpc [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    match_keys = Array(numrows, sint)
+    
+    match_bits = Matrix(numrows, args.n_bits, sint)
+    if args.sort_type == BIT_RADIX_SORT:
+        match_bits.assign_vector(
+            sint.get_input_from(0, size = numrows * args.n_bits))
+        match_keys.assign_vector(reconstruct(match_bits))
+    else:
+
+        match_keys.assign_vector(sint.get_input_from(0, size = numrows))
```
Comment (eriktaubeneck) [inline research-prototype/Programs/Source/erik.mpc:208]: these seem to consume different numbers of inputs

File: research-prototype/Programs/Source/erik.mpc [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+    reports = Matrix(numrows, 3, sint)
+    reports.assign_vector(sint.get_input_from(0, size = numrows * 3))
+
+    # Now do the sort
+    # Radix sort
+
+    if args.do_sort:
```
Comment (eriktaubeneck) [inline research-prototype/Programs/Source/erik.mpc:216]: is there any reason we wouldn't do the sort?

File: research-prototype/Programs/Source/erik.mpc [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            mask = breakdown == breakdown_key
+            # Sum up the final_credits for this breakdown
+            # We can do this with sum, but tree_reduce is more efficient.
+            breakdown_key_sums[breakdown_key] = tree_reduce(lambda x,y: x + y,
+                                                            mask * final_credits)
+            return breakdown_key_sums
+
+        return None
+
+def main():
```
Comment (eriktaubeneck) [inline research-prototype/Programs/Source/erik.mpc:185]: not sure if this is particularly helpful? it just forces everything to be indented and doesn't actually change the functionality

File: research-prototype/Programs/Source/erik.mpc [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+from pathlib import Path
+from Compiler import types, library, instructions, sorting
+
+VALUE = 0
+IS_TRIGGER = 1
+BREAKDOWN_KEY = 2
+
+RADIX_SORT = 0
+BIT_RADIX_SORT = 1
+BATCHER_SORT = 2
```
Comment (eriktaubeneck) [inline research-prototype/Programs/Source/erik.mpc:17]: this would be much better as strings

File: research-prototype/Programs/Source/erik.mpc [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+import argparse
+import os
+import sys
+from typing import Optional
+from pathlib import Path
+from Compiler import types, library, instructions, sorting
+
+VALUE = 0
+IS_TRIGGER = 1
+BREAKDOWN_KEY = 2
```
Comment (eriktaubeneck) [inline research-prototype/Programs/Source/erik.mpc:13]: why not continue to use the Enum.IntEnum?

**Review by eriktaubeneck — CHANGES_REQUESTED** [review]
Generally, there is A LOT going on here. It would be great if this was based off the existing `mp-spdz-prototype` branch, and edits were made directly to the ipae2e file, so that it was easy to see the differences. additionally, it would be great to have 1 PR for each incremental change, (i.e. one just focused on vectorization, and a different one with extra sort options, etc.)

File: research-prototype/Programs/Source/erik.mpc [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    return result
+
+def get_args(args) -> argparse.Namespace:
+    """
+    Get the arguments from the remainder of compile.py command line.
+    """
+
+    # It turns out that you *can* get arguments from the compile.py
+    # command.  The arguments are everything after the name
+    # of the file to be compiled.
```
[the remark this answers is not in the document]
Comment (eriktaubeneck) [reply research-prototype/Programs/Source/erik.mpc:45]: found another bug this this... calling `../../MP-SPDZ/compile.py -C -R 32 ipae2e 10` ends up naming the compiled programs `ipae2e-10`, which means you have to specify which version you're running...

**Pull request closed without merging**